server/item/recipe/recipe.go

```diff
@@ -0,0 +1,93 @@
+package recipe
+
+import (
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/df-mc/dragonfly/server/world"
+)
+
+// Recipe is implemented by all recipe types.
+type Recipe interface {
+	// Input returns the items required to craft the recipe.
+	Input() []item.Stack
+	// Output returns the items that are produced when the recipe is crafted.
+	Output() []item.Stack
+	// Block returns the block that is used to craft the recipe.
+	Block() world.Block
+	// Priority returns the priority of the recipe. Recipes with lower priority are preferred compared to recipes with
+	// higher priority.
+	Priority() uint32
+}
+
+// ShapelessRecipe is a recipe that has no particular shape.
+type ShapelessRecipe struct {
+	recipe
+}
+
+// NewShapelessRecipe creates a new shapeless recipe and returns it.
+func NewShapelessRecipe(input []item.Stack, output []item.Stack, block world.Block, priority uint32) ShapelessRecipe {
```

Comment:
This could also do with a better documentation. See comment on `NewShapedRecipe`.